drivers/gpio/gpio_tca9538.c

```diff
@@ -0,0 +1,601 @@
+			     sizeof(cmd), (uint8_t *)&input_states,
+			     sizeof(input_states));
+	if (ret == 0) {
+		drv_data->pin_state.input = input_states;
+	}
+	return ret;
+}
+
+/**
+ * @brief Handles interrupt triggered by the interrupt pin of TCA9538 I/O Port.
+ *
+ * If nint_gpios is configured in device tree then this will be triggered each
+ * time a gpio configured as an input changes state. The gpio input states are
+ * read in this function which clears the interrupt.
+ *
+ * @param arg Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 If successful.
+ * @retval Negative value for error code.
+ */
+static int gpio_tca9538_handle_interrupt(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	struct tca9538_drv_data *drv_data = dev->data;
+	struct tca9538_irq_state *irq_state = &drv_data->irq_state;
+	int ret = 0;
+	uint8_t previous_state;
+	uint8_t current_state;
+	uint8_t transitioned_pins;
+	uint8_t interrupt_mask = 0;
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	/* Store previous input state then read new value */
+	previous_state = drv_data->pin_state.input;
+	ret = update_input(dev);
+	if (ret != 0) {
+		goto out;
+	}
+
+	/* Find out which pins have changed state */
+	current_state = drv_data->pin_state.input;
+	transitioned_pins = previous_state ^ current_state;
+
+	/* Mask gpio transactions with rising/falling edge interrupt config */
+	interrupt_mask = (irq_state->interrupt_rising & transitioned_pins &
+			  current_state);
+	interrupt_mask |= (irq_state->interrupt_falling & transitioned_pins &
+			   previous_state);
+
+out:
+	k_sem_give(&drv_data->lock);
+
+	if (ret == 0) {
+		gpio_fire_callbacks(&drv_data->cb, dev, interrupt_mask);
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Work handler for TCA9538 interrupt
+ *
+ * @param work Work struct that contains pointer to interrupt handler function
+ */
+static void gpio_tca9538_work_handler(struct k_work *work)
+{
+	struct tca9538_drv_data *drv_data =
+		CONTAINER_OF(work, struct tca9538_drv_data, work);
+
+	gpio_tca9538_handle_interrupt((struct device *)drv_data->dev);
+}
+
+/**
+ * @brief ISR for intterupt pin of TCA9538
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param gpio_cb Pointer to callback function struct
+ * @param pins Bitmask of pins that triggered interrupt
+ */
+static void gpio_tca9538_init_cb(const struct device *dev,
+				 struct gpio_callback *gpio_cb, uint32_t pins)
+{
+	struct tca9538_drv_data *drv_data =
+		CONTAINER_OF(gpio_cb, struct tca9538_drv_data, gpio_cb);
+
+	ARG_UNUSED(pins);
+
+	k_work_submit(&drv_data->work);
+}
+
+/**
+ * @brief Initialization function of TCA9538
+ *
+ * This sets initial input/ output configuration, output states and
+ * input polarity. The interrupt is configured if this is enabled.
+ *
+ * @param dev Device struct
+ * @return 0 if successful, failed otherwise.
+ */
+static int gpio_tca9538_init(const struct device *dev)
```

Comment:
Traditionally we keep the init function towards the bottom of the driver file, just above the `struct` defining the api functions. Let's move it there.